src/flatten.c

```diff
@@ -11,7 +11,7 @@ const char* objtype(SEXP x) {
 
 SEXP flatten_impl(SEXP x) {
   if (TYPEOF(x) != VECSXP) {
-    stop_bad_type(x, "a list", NULL, ".x");
+    stop_bad_type(x, "a list", NULL, ".x", R_BaseEnv);
```

Comment:
Could vendor the rlang C library to call `r_peek_frame()`. Or reimplement it with the base lib:

```c
r_obj* r_peek_frame() {
  return r_eval(peek_frame_call, r_envs.empty);
}

  // Initialisation
  r_obj* current_frame_body = KEEP(r_parse_eval("as.call(list(sys.frame, -1))", r_envs.base));
  r_obj* current_frame_fn = KEEP(r_new_function(r_null, current_frame_body, r_envs.empty));
  peek_frame_call = r_call(current_frame_fn);
  r_preserve(peek_frame_call);
```

Reply:
In general I don't pass frames around and just peek the frame at throw site. This is what `r_abort()` does. This assumes that `.Call()` was called in the relevant user facing function.

If instead we need to forward a call, then I pass a `struct r_lazy` where `.x` either contains a frame environment (`.env` is NULL or the empty env, this represents a forced promise): https://github.com/r-lib/vctrs/blob/7c9d973b3c432da00329ae9a1873236cabd07abb/src/subscript-loc.c#L435

Or a symbol to be evaluated in a frame and that resolves to a `call` argument: https://github.com/r-lib/vctrs/blob/7c9d973b3c432da00329ae9a1873236cabd07abb/src/subscript-loc.c#L450

Reply:
How about I merge this PR (since it's massive) and then tackle the C stuff separately?